pt-lua.c

```diff
@@ -0,0 +1,966 @@
+  if(level == 0 || !lua_istable(L, -1))
+    return false;
+
+  lua_pushnil(L);  /* Initial key. */
+
+  while(lua_next(L, -2)) {
+    /* We are only interested in String keys. */
+    if(lua_type(L, -2) == LUA_TSTRING) {
+      /* Avoid "_G" recursion in global table. The global table is also part of
+         global table :). */
+      if(!strcmp(lua_tostring(L, -2), "_G")) {
+        /* Remove value and continue. */
+        lua_pop(L, 1);
+        continue;
+      }
+
+      /* Is it the function we are looking for? */
+      if(lua_rawequal(L, fn_idx, -1)) {
+        /* Remove value and keep name. */
+        lua_pop(L, 1);
+        return true;
+      }
+      /* If not go one level deeper and get the value recursively. */
+      else if(findfield(L, fn_idx, level - 1)) {
+        /* Remove the table but keep name. */
+        lua_remove(L, -2);
+
+        /* Add a "." in between. */
+        lua_pushliteral(L, ".");
+        lua_insert(L, -2);
+
+        /* Concatenate last 3 values, resulting "table.some_func". */
+        lua_concat(L, 3);
+
+        return true;
+      }
+    }
+
+    /* Pop the value. */
+    lua_pop(L, 1);
+  }
+
+  return false;
+}
+
+
+/* Pushes a function name if found in the global table and returns true.
+   Returns false otherwise. */
+/* Expects the function to be pushed in the stack. */
+static bool pushglobalfuncname(lua_State *L) {
+  int top = lua_gettop(L);
+
+  /* Start from the global table. */
+  lua_pushglobaltable(L);
+
+  if(findfield(L, top, 2)) {
+    lua_remove(L, -2);
+    return true;
+  }
+
+  lua_pop(L, 1);
+  return false;
+}
+
+
+/* Returns the maximum number of levels in Lua stack. */
+static int countlevels(lua_State *L) {
+  lua_Debug ar;
+  int li = 1, le = 1;
+
+  /* Find an upper bound */
+  while (lua_getstack(L, le, &ar)) {
+    li = le, le *= 2;
+  }
+
+  /* Do a binary search */
+  while (li < le) {
+    int m = (li + le) / 2;
+
+    if (lua_getstack(L, m, &ar)) li = m + 1;
+    else le = m;
+  }
+
+  return le - 1;
+}
+
+
+/* Counts the number of white and black frames in the Pallene call stack. */
+static void countframes(pt_fnstack_t *fnstack, int *mwhite, int *mblack) {
+  *mwhite = *mblack = 0;
+
+  for(int i = 0; i < fnstack->count; i++) {
+    *mwhite += (fnstack->stack[i].type == PALLENE_TRACER_FRAME_TYPE_C);
+    *mblack += (fnstack->stack[i].type == PALLENE_TRACER_FRAME_TYPE_LUA);
+  }
+}
+
+
+/* Responsible for printing and controlling some of the traceback fn parameters. */
+static void makeitcount(lua_State *L, luaL_Buffer *buf, bool *ellipsis, int *pframes, int nframes) {
+  /* We have printed the frame, even tho it might not be visible ;). */
+  (*pframes)++;
+
+  /* Should we print? Are we at any point in top or bottom printing threshold? */
+  bool should_print = (*pframes <= PT_LUA_TRACEBACK_TOP_THRESHOLD)
+    || ((nframes - *pframes) <= PT_LUA_TRACEBACK_BOTTOM_THRESHOLD);
+
+  if(luai_likely(should_print))
```

Comment:
No need to use luai_likely here. This macro has marginal benefit and only in the guts of the hottest of hot loops. Not the case here.